### ISSUE TITLE
bug: react custom elements does not account for dist-custom-elements-bundle

### ISSUE THREAD

**Issue by ianldgs** [ionic-team/stencil-ds-output-targets#136]
With the release of Stencil 2.4.0, we now have a new option for output target, the `dist-custom-elements`.
It will generate a bundle with each component on their own file: 

`class Component extends HTMLElement { ...compiled }`
`customElements.define('component', Component)`-ish

Some Stencil components, like layout components, can cause a lot of cumulative layout shift due to the lazy load.

It would be really nice if `react-output-target` had an option to import from the `dist-custom-elements`, so that bundlers can leverage three-shaking.

I think that it would be as simple as adding an option, and if `true`, add to the react output:

```diff
+import { Component as RawComponent } '${componentCorePackage}/components/component';
export const Component = /*@__PURE__*/createReactComponent<JSX.Component, HTMLComponentElement>('component');
```

The main challenge would be to make the `Component` variable use the `RawComponent` variable, so that static analysis can tell when to include it in the bundle or not.

Any thoughts?

PS: I'd be willing to contribute to a PR, once agreed on how to proceed.

**Comment by bjankord** [comment]
This would be great, for now our team has held off on the react output and I've considered writing my out output target code for React but would love a first class option from the stencil team instead.

**Comment by krail** [comment]
any progess on this?

**Comment by ianldgs** [comment]
@krail I have created #142 but have heard nothing whatsoever from the Stencil team. Tested it with npm workspaces and it works great. The only issue is polyfills. If you need them, you will need to load your entire application's bundle only after they are defined. Can probably be done with `@open-wc/polyfills-loader`.

**Comment by liamdebeasi** [comment]
Thanks for the issue. This has been resolved via https://github.com/ionic-team/stencil-ds-output-targets/pull/182 and will be available in an upcoming release of the React output target.

**Comment by Arvind6353** [comment]
@liamdebeasi Am trying to use the includeImportCustomElements option.

  **reactOutputTarget({
      componentCorePackage: 'fw-design-core',
      proxiesFile: '../react/src/components.ts',
      includeImportCustomElements: true,
      customElementsDir: '????',
    }),**
    
    What exactly should the customElementsDir point to ?
    
    
    My stencil.config.ts config as below :
    

**outputTargets: [
    {
      type: 'dist',
      esmLoaderPath: '../loader',
      copy: [{ src: 'assets' }],
    },
    {
      type: 'dist-custom-elements-bundle',
    },
    {
      type: 'docs-readme',
      footer: '*Built with Stencil',
    },
    {
      type: 'www',
      serviceWorker: null, // disable service workers
    },
    {
      type: 'custom',
      generator: generateJsonDocs,
      name: 'custom-element-docs',
    },

    reactOutputTarget({
      componentCorePackage: 'fw-design-core',
      proxiesFile: '../react/src/components.ts',
      includeImportCustomElements: true,
      customElementsDir: '????',
    }),**
    
    the docs says it will default to components dir . Am not sure where is the components directory present. Is it in the dist folder ?

**Comment by liamdebeasi** [comment]
`customElementsDir` is optional. It will default to the `components` directory if no value is provided. You only need to set `customElementsDir` if you set the `dir` option to something other than `components` in the `dist-custom-elements-bundle` object.

**Comment by Arvind6353** [comment]
@liamdebeasi By Default when we use dist-custom-elements-bundle in the output target option, 
As per stencil docs
**By default, the custom elements bundle will be written to dist/custom-elements/index.js. This directory can be configured using the output target's dir config.** 

https://stenciljs.com/docs/custom-elements

In this case should we by default set the dir to **components** for this dist-custom-elements-bundle object.?

**Comment by liamdebeasi** [comment]
Ah yes you are correct. I was looking at Ionic's config while writing that comment and we set it to `components` 🤦 

Let me take a closer look and I will follow up.

**Comment by Arvind6353** [comment]
@liamdebeasi  Ah okay. 

 import type { JSX } from 'fw-design-core/dist/custom-elements';

I also see that the below files are not present . 

dist/custom-elements folder contains only index.d.ts and index.js. so the below path is not found.

**import { FwButton as FwButtonCmp } from 'fw-design-core/dist/custom-elements/fw-button.js';
import { MyComponent as MyComponentCmp } from 'fw-design-core/dist/custom-elements/my-component.js';**

**Comment by Arvind6353** [comment]
@liamdebeasi  I guess the proxyfile should be 
`import { FwButton as FwButtonCmp } from 'fw-design-core/dist/custom-elements';
import { MyComponent as MyComponentCmp } from 'fw-design-core/dist/custom-elements';`

instead of 
`import { FwButton as FwButtonCmp } from 'fw-design-core/dist/custom-elements/fw-button.js';
import { MyComponent as MyComponentCmp } from 'fw-design-core/dist/custom-elements/my-component.js';`

The components are expored from the dist/custom-elements/index.js and no separate files are created for each component in the custom-elements bundle.

**Comment by liamdebeasi** [comment]
Yes, `dist-custom-elements-bundle` will generate one big file and `dist-custom-elements` will split out components into their own files. It sounds like we handle the second use case but not the first use case.

I can re-open this issue to address the `dist-custom-elements-bundle` use case.

**Comment by Arvind6353** [comment]
@liamdebeasi  Thank you. If I keep output target to **dist-custom-elements**, then **components folder** is created by default with individual components. **dist-custom-elements** output target config is not mentioned in the docs :(

**Comment by Arvind6353** [comment]
@liamdebeasi If we use dist-custom-elements / dist-customs-elements-bundle, we have the drawbacks mentioned in the docs right . 
1) Assets have to be manually moved 

2) This allows tree shaking but its good to avoid using defineCustomElements as mentioned below in the docs ?
**When defineCustomElements() is called, it will define every component in the bundle. However, it does not run automatically and it will not be called if it is not imported and executed. It can also result in a larger bundle size if there are unused components being imported.**

3) Polyfills are provided by default, if we need to support older browsers we need to manually add the polyfills. 
where as if we go with dist-lazy-load approach , we get these features baked in but tree shaking is not possible.


Please correct me if my understanding is wrong ?

**Comment by Arvind6353** [comment]
@liamdebeasi Since am using custom-elements , how do I import **setAssetPath** for handling local assets

```
import { defineCustomElements, setAssetPath } from 'my-library/dist/custom-elements';

setAssetPath(document.currentScript.src);
defineCustomElements();
```

In the docs its shown as being imported from **dist-custom-elements-bundle**. But here we are using **dist-custom-elements** and **setAssetPath** is not exported in this module

**Comment by splitinfinities** [comment]
Hey folks, just a heads up that we're going in the direction of recommending people to use `dist-custom-elements` with your bindings - not the `dist-custom-elements-bundle`. If there are use cases for this behavior, please share here, but it's the generalized recommendation to use `dist-custom-elements` esp. for the tree-shaking features your component libraries' consumers would like to take advantage of. 

With that said, unless we can get signal, we're going to focus our effort on improving the `dist-custom-elements` behavior when used with any supported bindings. We don't have a timeline on the `dist-custom-elements-bundle`, but would love to gather feedback or usage examples for this output target. 

Please add signal to #185

**Comment by Arvind6353** [comment]
> @liamdebeasi Am using dist-custom-elements , how do I import **setAssetPath** for handling local assets
> 
> ```
> import { defineCustomElements, setAssetPath } from 'my-library/dist/custom-elements';
> 
> setAssetPath(document.currentScript.src);
> defineCustomElements();
> ```
> 
> In the docs its shown as being imported from **dist-custom-elements-bundle**. But here we are using **dist-custom-elements** and **setAssetPath** is not exported in this module

@splitinfinities ,we have a bug with **dist-custom-elements** react binding. **setAssetPath** is not being exported from the react binding wrapper component.

**Comment by splitinfinities** [comment]
@Arvind6353 That seems to be a separate issue. Can you please file a separate bug for us to keep track of and prioritize?

**Comment by Arvind6353** [comment]
> @Arvind6353 That seems to be a separate issue. Can you please file a separate bug for us to keep track of and prioritize?

https://github.com/ionic-team/stencil-ds-output-targets/issues/186
Have raised a new one

**Comment by arvindanta** [comment]
Hi Team, Were you able to auto import the dependent components and register them as custom-elements without using defineCustomElements()?
